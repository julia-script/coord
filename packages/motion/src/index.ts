export * from "./context";
export * from "./player";
export * from "./utils";
export * from "./movie";
export * from "./flow";
export * from "./tweening";
export * from "./controls";
export * from "./effects";
