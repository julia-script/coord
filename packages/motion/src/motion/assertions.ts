import {
  YieldedType,
  isObject,
} from "@coord/core";
import {
  ContextRequest,
  EndRequest,
  MakeStateRequest,
  MotionRequest,
  MotionState,
} from "./types";

export const isRequest = (
  value: unknown
): value is MotionRequest => {
  return isObject(value) && "type" in value;
};

export const isContextRequest = <
  TState extends MotionState
>(
  value: unknown
): value is YieldedType<
  ContextRequest<TState>
> => {
  return (
    isRequest(value) &&
    value.type === "REQUEST_CONTEXT"
  );
};

export const isMakeStateRequest = <
  TKey extends string,
  TValue
>(
  value: unknown
): value is MakeStateRequest<TKey, TValue> => {
  return (
    isRequest(value) &&
    value.type === "MAKE_STATE" &&
    "state" in value &&
    isObject(value.state) &&
    "key" in value.state &&
    "initialState" in value.state
  );
};

export const isEndRequest = (
  value: unknown
): value is EndRequest => {
  return (
    isRequest(value) &&
    value.type === "REQUEST_END"
  );
};
