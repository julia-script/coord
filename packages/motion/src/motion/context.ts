import {
  isContextRequest,
  isEndRequest,
  isMakeStateRequest,
  isRequest,
} from "./assertions";
import {
  MotionBuilder,
  MotionMeta,
  MotionRequest,
  MotionScene,
  MotionSettings,
  MotionState,
  SceneState,
} from "./types";

type FullState<TState extends MotionState> = TState & {
  $transition: number;
  $frame: number;
  $done: boolean;
};
type FullStateKeys<TState extends MotionState> = keyof FullState<TState> &
  string;

export class MotionContext<TState extends MotionState> {
  _state: FullState<TState>;
  frames: FullState<TState>[] = [];
  meta: MotionMeta = {
    title: "Untitled",
    start: 0,
    duration: 0,
    scenes: {},
  };

  settings: MotionSettings = {
    fps: 60,
    physicsFps: 120,
  };

  get time() {
    return this.frames.length / this.settings.fps;
  }
  constructor() {
    this._state = {} as FullState<TState>;
  }

  static from<TScene extends MotionScene<MotionBuilder>>(
    scene: TScene,
    settings: Partial<MotionSettings> & { transitionDuration?: number } = {}
  ): MotionContext<SceneState<TScene>> {
    const context = new MotionContext<SceneState<TScene>>();
    context.meta.title = scene.meta.title;
    context.meta.description = scene.meta.description;

    const { transitionDuration = 0, ...rest } = settings;
    context.settings = { ...context.settings, ...rest };
    const builderIterator = scene.builder();

    let transitionDurationInFrames = context.settings.fps * transitionDuration;
    context._state.$done = false;

    let done = false;
    while (true) {
      if (done === false) {
        const currentIteration = builderIterator.next();
        done = !!currentIteration.done;
        if (isRequest(currentIteration.value)) {
          context.respondRequest(currentIteration.value);
          continue;
        }
      }
      if (
        done === true &&
        context.frames.length >= transitionDurationInFrames
      ) {
        break;
      }

      context._state.$frame = context.frames.length;

      context._state.$transition =
        transitionDurationInFrames <= 0
          ? 1
          : Math.min(
              (context.frames.length + 1) / transitionDurationInFrames,
              1
            );

      context.pushFrame();
    }
    if (context.frames.length === 0) {
      context._state.$done = true;
      context.pushFrame();
    }
    return context;
  }

  pushFrame() {
    this.meta.duration++;
    this.frames.push({
      ...this._state,
    });
  }

  setState<TKey extends FullStateKeys<TState>>(
    key: TKey,
    value: (typeof this._state)[TKey]
  ) {
    this._state[key] = value;
  }

  respondRequest<TRequest extends MotionRequest>(request: TRequest) {
    if (isContextRequest<TState>(request)) {
      request.context = this;
    } else if (
      isMakeStateRequest<
        FullStateKeys<TState>,
        FullState<TState>[FullStateKeys<TState>]
      >(request)
    ) {
      this.makeState(request.state.key, request.state.initialState);
    } else if (isEndRequest(request)) {
      this._state.$done = true;
    }
  }

  makeState<TKey extends FullStateKeys<TState>>(
    key: TKey,
    value: FullState<TState>[TKey]
  ) {
    this.setState(key, value);

    for (let i = this.frames.length - 1; i >= 0; i--) {
      const frame = this.frames[i]!;
      if (key in frame) {
        break;
      }
      Object.assign(frame, { [key]: value });
    }
  }
}
