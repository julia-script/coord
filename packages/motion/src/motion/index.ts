export * from "./assertions";
export * from "./context";
export * from "./requests";
export * from "./types";
export * from "./make-scene";
export * from "./run-motion";
