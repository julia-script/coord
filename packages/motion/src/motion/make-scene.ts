import { SceneMetaish, generateMeta } from "@/utils";
import { MotionBuilder } from "./types";

export function makeMotion<TBuilder extends MotionBuilder>(
  meta: SceneMetaish,
  builder: TBuilder
) {
  return {
    meta: generateMeta(meta),
    builder,
  };
}
