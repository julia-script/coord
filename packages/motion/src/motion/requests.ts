import { isFunction } from "@coord/core";
import { MotionState } from "./types";
import type { MotionContext } from "./context";
import { control } from "@/controls/control";

export function* requestContext<
  TState extends MotionState
>() {
  const request: {
    type: "REQUEST_CONTEXT";
    context?: MotionContext<TState>;
  } = {
    type: "REQUEST_CONTEXT",
  };

  yield request;

  if (!request.context) {
    throw new Error("Context is not provided");
  }

  return request.context;
}

export function* requestEnd() {
  yield { type: "REQUEST_END" } as const;
}

export function* makeState<
  TKey extends string,
  TValue
>(key: TKey, initialState: TValue) {
  const context = yield* requestContext();
  yield {
    type: "MAKE_STATE",
    state: { key, initialState },
  } as const;

  const getter = () =>
    context._state[key] as TValue;
  const setter = (
    value: TValue | ((prev: TValue) => TValue)
  ) => {
    context._state[key] = isFunction(value)
      ? value(getter())
      : value;
  };
  return control(getter, setter);
}
