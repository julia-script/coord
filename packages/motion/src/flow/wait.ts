import { requestContext, requestPassTime } from "@/context";
import { countFrames } from "@/utils";

export function* wait(duration: number) {
  const context = yield* requestContext();
  const fps = context.settings.fps;
  const [frames, remainder] = countFrames(context.time, fps, duration);

  for (let i = 0; i < frames; i++) {
    yield;
  }

  yield* requestPassTime(remainder);
}
