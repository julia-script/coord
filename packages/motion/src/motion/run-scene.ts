import { MotionContext } from "./context";
import { MotionBuilder, MotionScene, MotionSettings } from "./types";

export function runMotion<TScene extends MotionScene<MotionBuilder>>(
  scene: TScene,
  settings?: Partial<MotionSettings> & { transitionDuration?: number }
) {
  const context = MotionContext.from(scene, settings);

  return {
    _context: context,
    frames: context.frames,
    meta: context.meta,
    settings: context.settings,
  };
}
