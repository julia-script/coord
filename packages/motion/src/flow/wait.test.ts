import { makeState, runMotion } from "@/motion";
import { makeMotion } from "@/motion";
import { test, describe, expect } from "vitest";
import { wait } from "./wait";

describe("wait", async () => {
  test("waits the specified time before continuing", async () => {
    const scene = makeMotion(
      "Test",
      function* () {
        const a = yield* makeState(
          "a",
          "Waiting"
        );
        yield* wait(1);
        yield a.set("Done");
      }
    );
    const executed = runMotion(scene, {
      fps: 3,
    });

    expect(executed.frames).toMatchSnapshot();
  });
});
