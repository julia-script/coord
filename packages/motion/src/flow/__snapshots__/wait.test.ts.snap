// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wait > waits the specified time before continuing 1`] = `
[
  {
    "$done": false,
    "$frame": 0,
    "$transition": 1,
    "a": "Waiting",
  },
  {
    "$done": false,
    "$frame": 1,
    "$transition": 1,
    "a": "Waiting",
  },
  {
    "$done": false,
    "$frame": 2,
    "$transition": 1,
    "a": "Waiting",
  },
  {
    "$done": false,
    "$frame": 3,
    "$transition": 1,
    "a": "Done",
  },
]
`;
